### PR TITLE
Migrate from RAFT to CUVS

### PR DESCRIPTION
**Pull request by tarang-jain**
Remove the dependency on `raft::compiled` and modify GPU implementations to use cuVS backend in place of RAFT.

A deeper insight into the dependency:
FAISS gets the ANN algorithm implementations such as IVF-Flat and IVF-PQ from cuVS. RAFT is meant to be a lightweight C++ header-only template library that cuVS relies on for the more fundamental / low-level utilities. Some examples of these are RAFT's device mdarray and mdspan objects; the RAFT resource object (`raft::resource`) that takes care of the stream ordering of device functions; linear algebra functions such as mapping, reduction, BLAS routines etc. A lot of the cuVS functions take the RAFT mdspan objects as arguments (for example `raft::device_matrix_view`). Therefore FAISS relies on both cuVS and RAFT. FAISS gets RAFT headers through cuVS and uses them to create the function arguments that can be consumed by cuVS. Note that we are not explicitly linking FAISS against `raft::raft` or `raft::compiled`. Only the required headers are included and compiled rather than compiling the whole RAFT shared library. This is the reason we still see mentions of `raft` in FAISS.

### CODE REVIEW

**Comment by asadoughi** [comment]
Hi @tarang-jain! Thanks for the PR. I quickly skimmed through the PR and noticed there are still a lot of references to RAFT. Is the plan to completely remove all references to RAFT and replace with cuVS? For example, try `grep -r -nH -i raft .` in your local checkout.

**Comment by tarang-jain** [comment]
@asadoughi thanks for your comment. I have updated the PR description!

File: INSTALL.md
Comment (asadoughi) [inline INSTALL.md:113]: What about lines 9, 20-21, 27, 38-39? Will we have a faiss-gpu-cuvs package?

File: benchs/bench_ivfpq_cuvs.py
Comment (asadoughi) [inline benchs/bench_ivfpq_cuvs.py:44]: Have you run this before and after the change to determine the change (if any) in performance?

File: benchs/bench_ivfflat_cuvs.py
Comment (asadoughi) [inline benchs/bench_ivfflat_cuvs.py:46]: Have you run this before and after the change to determine the change (if any) in performance?

File: faiss/gpu/CMakeLists.txt
Comment (asadoughi) [inline faiss/gpu/CMakeLists.txt:293]: Is this still necessary with cuvs?

File: faiss/gpu/utils/CuvsUtils.h
Comment (asadoughi) [inline faiss/gpu/utils/CuvsUtils.h:56]: Should this be a CUVS error message?

File: faiss/gpu/test/TestGpuDistance.cu
Comment (asadoughi) [inline faiss/gpu/test/TestGpuDistance.cu:51]: "using cuvs"?

File: faiss/gpu/test/TestGpuIndexIVFFlat.cpp
Comment (asadoughi) [inline faiss/gpu/test/TestGpuIndexIVFFlat.cpp:334]: Commented out?

File: faiss/gpu/impl/CuvsCagra.cu
Comment (asadoughi) [inline faiss/gpu/impl/CuvsCagra.cu:188]: Dead code?

**Review by asadoughi — CHANGES_REQUESTED** [review]
Hi @tarang-jain - thanks for the changes. I provided some comments inline.

File: faiss/gpu/CMakeLists.txt
Comment (tarang-jain) [reply faiss/gpu/CMakeLists.txt:293]: Yes that is still needed to avoid any errors due to duplicate CUDA symbols.

File: faiss/gpu/utils/CuvsUtils.h
Comment (tarang-jain) [reply faiss/gpu/utils/CuvsUtils.h:56]: We currently use `RAFT_FAIL` as an error generator in cuVS too. We currently don't have a cuVS error message but I have marked it as a TODO. This certainly needs to be done to avoid confusion since it is a cuVS failure and not a RAFT failure.

File: faiss/gpu/test/TestGpuIndexIVFFlat.cpp
Comment (tarang-jain) [reply faiss/gpu/test/TestGpuIndexIVFFlat.cpp:334]: Updated.

File: faiss/gpu/impl/CuvsCagra.cu
Comment (tarang-jain) [reply faiss/gpu/impl/CuvsCagra.cu]: removed!

File: INSTALL.md
Comment (tarang-jain) [reply INSTALL.md:113]: I intentionally did not make changes to the docs (readme and install) because I thought that is something entirely under FAISS' governance. Don't want to modify that without approval from you guys. But to answer your question: yes it will be called `faiss-gpu-cuvs`.

Comment (asadoughi) [reply INSTALL.md:113]: Those lines reference packages distributed by the `nvidia` channel so if you have `faiss-gpu-cuvs` prepared we can coordinate the release.

**Comment by asadoughi** [comment]
Hi @tarang-jain - it looks like the test fails on https://github.com/facebookresearch/faiss/actions/runs/10587022390/job/29337315242#step:3:2446 related to a core dump. Are you able to debug the issue?

**Comment by asadoughi** [comment]
Regarding the segfault, consider rebasing off of the latest main to see if #3806 has an effect.

**Comment by tarang-jain** [comment]
@asadoughi I was able to resolve the seg fault. It was related to some bugs in the cuvs parts of bfKnn.

File: faiss/gpu/StandardGpuResources.cpp
Comment (tarang-jain) [reply faiss/gpu/StandardGpuResources.cpp:261]: Done!

File: CMakeLists.txt
Comment (asadoughi) [inline CMakeLists.txt:83]: Is RAFT still needed for compilation or we need to install rmm explicitly? 

https://github.com/facebookresearch/faiss/actions/runs/10889999572/job/30218385769?pr=3549
```
The link interface of target "cuvs::cuvs" contains:

    rmm::rmm

  but the target was not found.  Possible reasons include:
  ```

Comment (tarang-jain) [reply CMakeLists.txt:83]: @asadoughi I tried linking `raft::raft` but now I see another failure in ci https://github.com/facebookresearch/faiss/actions/runs/10908718713/job/30275337629?pr=3549

Comment (tarang-jain) [reply CMakeLists.txt:83]: rmm and raft should ideally be available through cuvs. But I have also tried explicitly linking `raft::raft`.

File: .github/actions/build_cmake/action.yml
Comment (asadoughi) [inline .github/actions/build_cmake/action.yml:62]: Perhaps try keeping the conda install of libraft here?

**Comment by asadoughi** [comment]
It looks like one test is still failing `TestTorchUtilsGPU.test_train_add_with_ids`. Have you had a chance to troubleshoot?

**Comment by tarang-jain** [comment]
@asadoughi yes I have been looking into it. It is quite strange. The following order of operations gives erroneous results:
`create a GPU torch tensor --> train GPU IVF-Flat Index --> search the index --> reset the index --> train again with the same tensor but this time have the data on CPU --> train the GPU IVF-Flat index --> search the index.`

However, if both the tensors are on the same device (either CPU or GPU), there is no failure and it works fine.

**Comment by asadoughi** [comment]
I tried addressing the current error message `Error while loading conda entry point: conda-libmamba-solver (module 'libmambapy' has no attribute 'QueryFormat')`  in #3904 by pinning the libmamba and libmambapy installations to `<2`, but that led to [another error with rapids-cmake](https://github.com/facebookresearch/faiss/actions/runs/11147695260/job/30983363857?pr=3904). Have you seen this error with rapids-cmake before?

**Comment by tarang-jain** [comment]
@asadoughi the error you pointed out is in the RAFT build. Not sure if it is relevant to this PR.

**Comment by asadoughi** [comment]
It looks like this PR is getting the same error now as well in the latest [run](https://github.com/facebookresearch/faiss/actions/runs/11151478626/job/30995601079?pr=3549): failing on this [line](https://github.com/rapidsai/rapids-cmake/blob/branch-24.10/RAPIDS.cmake#L93).

**Comment by tarang-jain** [comment]
Yes, I asked Robert, our CMake expert to look at this. I think he is busy with the 24.10 release deadlines, but I'm hoping he can give us a fix in the next few days.

**Comment by asadoughi** [comment]
Hi @tarang-jain - any update or ETA on the CMake issue? This is also causing failures on the `main` branch for FAISS (e.g. blocking PRs) so if we don't have an ETA to a fix I am leaning towards disabling the RAFT CI on `main` and it can be re-instated with this cuVS PR. Thanks!

File: faiss/gpu/GpuDistance.cu
Comment (benfred) [inline faiss/gpu/GpuDistance.cu:236]: I believe we can have mixed layout with cuvs now - this comment might be out of date

File: tests/CMakeLists.txt
Comment (benfred) [inline tests/CMakeLists.txt:125]:
```suggestion
```

**Comment by tarang-jain** [comment]
@asadoughi can the system glibc version be updated for running the cuvs github action? The cmake seg fault went away when I downgrade to cmake 3.26.4, but I see this error https://github.com/facebookresearch/faiss/actions/runs/11239339471/job/31246908895?pr=3549 which might mean a libc version mismatch.

**Comment by asadoughi** [comment]
AFAICT, [this CI](https://github.com/facebookresearch/faiss/blob/2e6551ffa3f6fbdb1ba814c2c531fb399b00d4e3/.github/workflows/build.yml#L42) is running on ubuntu-latest so you can try amending the shell commands in build_cmake/action.yaml to run `ldd --version` and the relevant `sudo apt-get -y install` command to install the specific glibc version.

File: faiss/gpu/test/CMakeLists.txt
Comment (asadoughi) [inline faiss/gpu/test/CMakeLists.txt:29]: Looks like there is an extra `>` here to be removed: x86_64-conda-linux-gnu-c++: error: unrecognized command-line option '-Xcompiler=-fopenmp>'

**Review by asadoughi — COMMENTED** [review]
Thanks! Looks like unpinned CMake with 24.08 works.

In addition to the torch CPU tensor issue, let's track
- migration to libcuvs 24.10, pending your discussion with @cjnolet 
  - including cuVS build time 1+ hr more than other CI jobs 
- cuVS in FAISS documentation - already filed #3823

File: faiss/gpu/test/torch_test_contrib_gpu.py
Comment (tarang-jain) [reply faiss/gpu/test/torch_test_contrib_gpu.py:85]: Done

**Comment by tarang-jain** [comment]
@asadoughi @cjnolet, due to the segmentation fault in the cagra tests with cuvs 24.10, I have reverted to cuVS 24.08. I will continue to work on getting 24.10 in on the side. But it is crucial to get this PR merged quickly.

**Comment by asadoughi** [comment]
@tarang-jain Did you mean to push those latest 4 commits to this PR? If we're rolling forward with 24.08 we should probably go back to [3e056ed](https://github.com/facebookresearch/faiss/pull/3549/commits/3e056ed911e73c998794de0342f7662dc5db3c01).

**Comment by tarang-jain** [comment]
@asadoughi I have fixed the segmentation fault occurring with 24.10 in the CAGRA tests and so we are very close to getting cuVS 24.10 in. There are still some `std::badalloc` in the torch CPU tests. If I am not able to fix those by EOD today, I will open a follow-up PR with 24.10 and revert this PR to https://github.com/facebookresearch/faiss/pull/3549/commits/3e056ed911e73c998794de0342f7662dc5db3c01

**Comment by tarang-jain** [comment]
I have downgraded cuVS to 24.08 again. I'm working on https://github.com/facebookresearch/faiss/pull/4021 to get 24.10 in after this PR is merged.

**Review by asadoughi — CHANGES_REQUESTED** [review]
Hi @tarang-jain! Thanks for updating. I have just a couple of suggestions inline.

File: .github/actions/build_cmake/action.yml
Comment (asadoughi) [reply .github/actions/build_cmake/action.yml:35]: Or, do we need it at all for cuVS 24.08? I thought we only needed it to be pinned for cuVS 24.10.

Comment (tarang-jain) [reply .github/actions/build_cmake/action.yml:35]: yes I can remove the cmake pinning. Some debugging changes creeped in during the constant back and forth.

File: faiss/gpu/GpuIndex.cu
Comment (tarang-jain) [reply faiss/gpu/GpuIndex.cu:46]: No, that was intentional. FAISS'  helper `getDeviceProperties` cache's the device major version internally, so we should be re-using that API.

Comment (asadoughi) [reply faiss/gpu/GpuIndex.cu:46]: Nice catch!

**Comment by tarang-jain** [comment]
@asadoughi what are the facebook internal checks that are failing?

**Pull request closed without merging**
